### PR TITLE
Developing Hallucination Guardrails

### PR DESCRIPTION
**Pull request by royziv11**
## Summary

Briefly describe the changes and the goal of this PR. Make sure the PR title summarizes the changes effectively.

## Motivation

Why are these changes necessary? How do they improve the cookbook?

---

## For new content

When contributing new content, read through our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md), and mark the following action items as completed:

- [ ] I have added a new entry in [registry.yaml](https://github.com/openai/openai-cookbook/blob/main/registry.yaml) (and, optionally, in [authors.yaml](https://github.com/openai/openai-cookbook/blob/main/authors.yaml)) so that my content renders on the cookbook website.
- [ ] I have conducted a self-review of my content based on the [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md#rubric):
  - [ ] Relevance: This content is related to building with OpenAI technologies and is useful to others.
  - [ ] Uniqueness: I have searched for related examples in the OpenAI Cookbook, and verified that my content offers new insights or unique information compared to existing documentation.
  - [ ] Spelling and Grammar: I have checked for spelling or grammatical mistakes.
  - [ ] Clarity: I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [ ] Correctness: The information I include is correct and all of my code executes successfully.
  - [ ] Completeness: I have explained everything fully, including all necessary references and citations.

We will rate each of these areas on a scale from 1 to 4, and will only accept contributions that score 3 or higher on all areas. Refer to our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md) for more details.

### CODE REVIEW

File: examples/Developing_hallucination_guardrails.ipynb [hunk 1]
```diff
@@ -0,0 +1,1101 @@
...
+   "execution_count": 1,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-05-20T21:05:07.490991Z",
+     "start_time": "2024-05-20T21:05:07.475169Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import json\n",
```
Comment (shyamal-anadkat) [inline examples/Developing_hallucination_guardrails.ipynb:28]: nit - can you isort these imports?

File: examples/Developing_hallucination_guardrails.ipynb [hunk 2]
```diff
@@ -0,0 +1,1101 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Developing Hallucination Guardrails\n",
+    "\n",
+    "In this notebook we'll walk through the process of developing a guardrail that checks model outputs and against hallucinations. \n",
```
Comment (shyamal-anadkat) [inline examples/Developing_hallucination_guardrails.ipynb:9]: can we briefly explain what a guardrail is and/or link to the existing cookbook: https://cookbook.openai.com/examples/how_to_use_guardrails?

Comment (shyamal-anadkat) [inline examples/Developing_hallucination_guardrails.ipynb:9]: check model outputs "for hallucinations"?

File: examples/Developing_hallucination_guardrails.ipynb [hunk 3]
```diff
@@ -0,0 +1,1101 @@
...
+    "\n",
+    "This notebook will focus on:\n",
+    "1. Building out a strong eval set\n",
+    "2. Identifying specific criteria to measure hallucinations\n",
+    "3. Improving accuracy with few-shot prompting"
```
Comment (shyamal-anadkat) [inline examples/Developing_hallucination_guardrails.ipynb:14]: improving accuracy of what? (can we make that bit more clear?)

File: examples/Developing_hallucination_guardrails.ipynb [hunk 4]
```diff
@@ -0,0 +1,1101 @@
...
+    "\n",
+    "1. Provide very descriptive metrics to evaluate whether a response is accurate\n",
+    "- It is important to break down this idea of \"truth\" in easily identifiable metrics that we can measure\n",
+    "- Metrics like truthfulness and relevance are difficult to measure. Giving concrete ways to score the statement can result in a more accurate guardrail\n",
+    "2. Ensure consistency across key terminology\n",
+    "- It is important to keep relevant terms such as KB articles, assistants, and users consistent across the prompt\n",
+    "- If we begin to use phrases such as assistant vs agent, the model could get confused\n",
+    "3. Start with the most advanced model\n",
+    "- GPT-4-Turbo may be the most expensive model, but it important to start with the most advanced so we can ensure a high degree of accuracy\n",
+    "- Once we have thoroughly tested out the guardrail and are confident in its performance, we can look to reducing cost by tuning it down to a 3.5 model\n",
```
Comment (shyamal-anadkat) [inline examples/Developing_hallucination_guardrails.ipynb:611]: a 3.5 model -> "gpt-3.5-turbo"

**Comment by royziv11** [comment]
@shyamal-anadkat - Made all of the necessary tweaks. Let me know if you have an additional suggestions. Thank you!

**Review by shyamal-anadkat — APPROVED** [review]
lgtm!

File: registry.yaml [hunk 5]
```diff
@@ -483,6 +483,14 @@
   tags:
     - completions
 
+- title: Developing Hallucination Guardrails
+  path: examples/Developing_hallucination_guardrails.ipynb
+  date: 2024-05-29
+  authors:
+    - royziv11
+  tags:
+    - completions
```
Comment (shyamal-anadkat) [inline registry.yaml:492]: we should create a tag for guardrails

Comment (royziv11) [reply registry.yaml:492]: Do I just need to add a guardrail tag here and for the other guardrail cookbook or is there somewhere else where we store the list of tags?

**Pull request merged**